Fix for DDC-3697 and DDC-3701

mpdude opened the pull request:
This is about the DQL parser generating wrong SQL because conditions may end in the "ON" clause of a LEFT JOIN instead of the WHERE clause (http://www.doctrine-project.org/jira/browse/DDC-3697).

It also fixes that the parser may be too sloppy on expected tokens, for example accepting a WHERE instead of WITH (http://www.doctrine-project.org/jira/browse/DDC-3701) or `identifier`s with backslashes like in `SELECT \foo.bar\baz FROM...`. The first problem probably would have surfaced earlier with this strict check in place.

mpdude committed:
Fix for DDC-3697

Also fix Lexer::match() so it does not accept T_WHERE when T_WITH is supposed to be matched. (DDC-3701)

mpdude commented:
@guilhermeblanco You've put many efforts into the parts touched by this PR – what do you think about it?

mpdude commented:
@Ocramius How can I help here to get this along?

mpdude committed:
Add tests for DDC-3701 (GH #1389)

mpdude commented:
Closed #1389 in favor of this PR as the problem is solved here. Added some tests for the #1389 problem.

mpdude commented:
Tests fail due to a change in doctrine/cache, see #1510

Ocramius commented:
@mpdude since #1510 was merged, rebasing this should bring the tests back to green

Ocramius commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -164,7 +164,7 @@ protected function getType(&$value)
 
             // Recognize identifiers
-            case (ctype_alpha($value[0]) || $value[0] === '_'):
+            case (ctype_alpha($value[0]) || $value[0] === '_' || $value[0] === '\\'):
```
There are no constants that start with `T__` or `T_\\` (the last one is quite un-possible).

I suggest adding more cases for those specifically instead

Ocramius replied:
Could need a separate lexer test as well

Ocramius replied:
After looking more carefully at [DDC-3701](http://www.doctrine-project.org/jira/browse/DDC-3701), I don't think that this fix is needed. In DQL, an identifier is basically anything "referenceable" (alias or entity name). An entity name cannot start with `\`, because DQL has no namespacing.

mpdude replied:
Ouch. I was not aware of that and in fact it *was* possible to use a FQCN with a leading `\` with no problems – probably due to the parser inaccuracy addressed here. 

Just change one of the tests in `SelectSqlGenerationTest` on `master` and see yourself :-).

mpdude replied:
And also have a look [here](https://github.com/doctrine/doctrine2/blob/master/tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php#L465) where a FQCN with a leading `\` is used in INSTANCE OF.

In the BCNF, both `InstanceOfParameter` and `RangeVariableDeclaration` build on `AbstractSchemaName`, which in turn is an `identifier`. `identifier` itself is a little... underspecified?

Ocramius replied:
@mpdude if `\` already exists in our docs, then agree, it's needed for consistency.

mpdude replied:
Adding a `\Doctrine\Tests\ORM\Query\LexerTest::testScannerRecognizesIdentifier` test that also covers leading `\` in FQCNs; plus a new \Doctrine\Tests\ORM\Query\ParserTest that asserts `AbstractSchemaName` from BCNF will have leading `\` stripped.

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -312,10 +312,9 @@ public function match($token)
 
         // short-circuit on first condition, usually types match
-        if ($lookaheadType !== $token && $token !== Lexer::T_IDENTIFIER && $lookaheadType <= Lexer::T_IDENTIFIER) {
+        if ($lookaheadType !== $token && ($token !== Lexer::T_IDENTIFIER || $lookaheadType <= Lexer::T_IDENTIFIER)) {
```
Are there specific parser tests for this? Could need some explicit coverage for these changes, if possible (if not, simply do tell: don't need you to write a complex test for no good reason)

mpdude replied:
Added some.

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -1678,8 +1677,6 @@ public function Join()
         if ($joinDeclaration instanceof AST\RangeVariableDeclaration) {
             $joinDeclaration->isRoot = false;
-
-            $adhocConditions = true;
```
Same as above

mpdude replied:
Not sure how I could reasonably cover this in a test.

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -1125,15 +1125,17 @@ public function walkJoin($join)
+                if ($join->conditionalExpression) {
+                    $conditions[] = '(' . $this->walkConditionalExpression($join->conditionalExpression) . ')';
+                }
                 $condExprConjunction = ($class->isInheritanceTypeJoined() && $joinType != AST\Join::JOIN_TYPE_LEFT && $joinType != AST\Join::JOIN_TYPE_LEFTOUTER)
```
Newline before this line

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -1148,7 +1150,10 @@ public function walkJoin($join)
                 }
 
-                $sql .= $condExprConjunction . implode(' AND ', $conditions);
+                if ($conditions) {
```
Can probably skip this `if`, as an empty array would yield the same result

mpdude replied:
No, because you'd add the `$condExprConjunction` to the `$sql` (an `AND` or `ON`) even when no actual condition would follow.

Ocramius replied:
Makes sense, thanks!

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -84,6 +84,23 @@ public function assertInvalidSqlGeneration($dqlToBeTested, $expectedException, a
         $this->fail($sql);
     }
 
+    public function testDDC3697WithPartialLoad()
```
I would add (to the method name) something that indicates that `WHERE` and `WITH` both work

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -84,6 +84,23 @@ public function assertInvalidSqlGeneration($dqlToBeTested, $expectedException, a
+        );
+    }
+
+    public function testDDC3697WithoutPartialLoad()
```
I would add (to the method name) something that indicates that `WHERE` and `WITH` both work

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -2242,6 +2259,43 @@ public function testHavingSupportResultVariableInAggregateFunction()
+     */
+    public function testFROMClauseAcceptsFQCNWithLeadingBackslash()
+    {
+        $this->_em->createQuery(
```
This test lacks an assertion. If you are just verifying that valid SQL is generated, then `$this->assertInternalType('string'), $sql)` is enough

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -2242,6 +2259,43 @@ public function testHavingSupportResultVariableInAggregateFunction()
+     */
+    public function testParserDoesNotAcceptWHEREWhenWITHIsExpected()
+    {
+        $this->_em->createQuery('
```
Same as above

mpdude replied:
No, this one is an `@expectedException`.

Ocramius replied:
Can you make it explicit via `setExpectedException` then?

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -2242,6 +2259,43 @@ public function testHavingSupportResultVariableInAggregateFunction()
+     */
+    public function testParserAcceptsJoinWithRangeVariableDeclatationAndNoWITHClause()
+    {
+        $this->_em->createQuery('
```
Same as above

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -2242,6 +2259,43 @@ public function testHavingSupportResultVariableInAggregateFunction()
+    public function testFROMClauseAcceptsFQCNWithLeadingBackslash()
+    {
+        $this->_em->createQuery(
+            'SELECT c FROM \Doctrine\Tests\Models\Company\CompanyContract c'
```
This is un-wanted. In DQL, you always use FQCNs, therefore the leading backslash should never match.

mpdude replied:
Then I'll remove that test.

Ocramius replied:
Since you said above that `\` exists in some parts of DQL, it's good to have consistency and support it everywhere. Sorry for the review mistake.

mpdude replied:
Wait – so you're saying we should allow the leading backslash?

mpdude replied:
What I was trying to say was that it *seems* that for the rest of the code the leading `\` makes no difference. And that there are some tests actually using it, but maybe the authors would have noticed and changed that *if* the parser had actually complained.

Ocramius replied:
If the leading backslash is allowed somewhere in DQL, then it should be allowed anywhere

mpdude committed:
Undisputed GH feedback

mpdude committed:
Make expected exception explicit

mpdude committed:
Improve Lexer tests, also asserting that identifiers can start with a backslash.

mpdude committed:
For consistency, remove leading backslashes on identifiers here.

Note that we're now "officially" supporting them, but tests for that go elsewhere.

mpdude committed:
Move tests from SelectSqlGenerationTest to LanguageRecognitionTest; add ParserTest

mpdude committed:
Add tests for Parser/Lexer fixes

mpdude committed:
Try to better name tests

I still feel the need for a comment, though

mpdude commented:
@Ocramius tried to get as much as possible of your feedback into this. Would be great if you could have another look at it.

mpdude commented:
After a few hours of sleep I come to recognize that this way of changing T_IDENTIFIER cannot be the right solution, as if would for example allow expressions like "SELECT \u FROM ... u" or "SELECT u.\someField FROM ..." as well.

Do we really want to relax the DQL grammar and explicitly allow for leading `\`s? This may take some work to get it right.

Ocramius commented:
No, that seems wrong indeed...
On Sep 20, 2015 17:24, "Matthias Pigulla" <notifications@github.com> wrote:

> After a few hours of sleep I come to recognize that this way of changing
> T_IDENTIFIER cannot be the right solution, as if would for example allow
> expressions like "SELECT \u FROM ... u" or "SELECT u.\someField FROM ..."
> as well.
>
> Do we really want to relax the DQL grammar and explicitly allow for
> leading \s? This may take some work to get it right.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/1395#issuecomment-141761870>.
>

guilhermeblanco commented:
@Ocramius @mpdude I should be able to review this during this week... =)

mpdude commented:
Support for leading `\` in class names (and later stripping it) has been added in 0424d870999c9fb0a044a8b5773f250376d2368f as a fix for [DDC-675](http://www.doctrine-project.org/jira/browse/DDC-675). At a first glance it looks as if the lexer should have choked anyway because a `\` was not allowed in the identifier preg back then.

mpdude commented:
The problem is that \Foo, Foo:Bar:Baz, Foo\Bar, Foo:\Bar etc. all match the [regexp in the Lexer](https://github.com/doctrine/doctrine2/blame/master/lib/Doctrine/ORM/Query/Lexer.php#L129) and are considered valid `identifier`s. `identifier` is used for the `AbstractSchemaName`, `AliasIdentificationVariable`, `IdentificationVariable` and so on.

So, if we want to catch "SELECT \u.foo:bar FROM Bar:Baz:\Qaz \u" at the parser level, we'd probably need to define `identifier` as `[a-z_][a-z0-9_]*` and have T_COLON and T_BACKSLASH types in the lexer.

Then

```
     AbstractSchemaName ::= FullyQualifiedClassName | NamespaceAliasedClassName
     FullyQualifiedClassName ::= ["\"] identifier {"\" identifier}*
     NamespaceAliasedClassName ::= identifier ":" identifier
```

But, as I see it, the colon prevents us from having the handy T_INPUT_PARAMETER token and we'd need to add some production rule for parameters as well... 

As I see it this is already broken right now (nothing caused by this PR), so we could also just merge this and walk away as if nothing had happened.

mpdude committed:
Make sure the Lexer and Parser discriminate identifiers, qualified names and aliased names

For example, "SELECT u.\foo\bar ..." should be rejected because only an identifier but not a qualified name must be used.

mpdude commented:
I think I have fixed the issue of allowing backslashes (and colons) in the wrong places by making T_IDENTIFIER accept [a-z_][a-z0-9_]* only. For qualified and aliased names, two new token types are added. 

Also added tests that hopefully illustrate it.

@Ocramius and @guilhermeblanco what do you guys think?

mpdude commented:
Can I do anything to make the review easier for you?

Ocramius commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -49,6 +49,8 @@ class Lexer extends \Doctrine\Common\Lexer
     const T_OPEN_CURLY_BRACE    = 18;
     const T_CLOSE_CURLY_BRACE   = 19;
+    const T_ALIASED_NAME        = 20;
+    const T_QUALIFIED_NAME      = 21;
```
Probably "fully qualified"

Ocramius commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -175,7 +178,13 @@ protected function getType(&$value)
-                return self::T_IDENTIFIER;
+                if (strpos($value, '\\') !== false) {
+                    return self::T_QUALIFIED_NAME;
+                } else if (strpos($value, ':') !== false) {
```
Since you use early returns, there is no need for `else`

Ocramius replied:
Also, shouldn't this condition come first, since any name containing a `:` is an alias?

mpdude replied:
Question is whether "SomeAlias:Some\Sub\Class" should be allowed or not?

Ocramius replied:
@mpdude no `SomeAlias:Some\Sub\Class` is absolutely a bad idea :-) Anyway, we should really kill aliases at some point.

mpdude replied:
Then it is ok that way: If there is an `\` in it, it is *not* an alias (will lead to unresolvable class in the "Alias:Sub\Class" case). Aliases only work for "Foo:Bar" like names.

Ocramius commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -175,7 +178,13 @@ protected function getType(&$value)
+                    return self::T_QUALIFIED_NAME;
+                } else if (strpos($value, ':') !== false) {
+                    return self::T_ALIASED_NAME;
+                } else {
```
Same as above

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -311,11 +311,15 @@ public function match($token)
+        if ($lookaheadType !== $token && ($token !== Lexer::T_IDENTIFIER || $lookaheadType <= Lexer::T_IDENTIFIER)) {
             $this->syntaxError($this->lexer->getLiteral($token));
         }
-
```
Revert (we keep newlines between logical blocks ;-) )

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -311,11 +311,15 @@ public function match($token)
+         * - If expectation is to get T_IDENTIFIER and we find one of the tokens that are reserved words *but* would work as identifiers as well (e. g. "FROM", DDC-505) -> ok.
+         * - Else fail.
+         */
+        if ($lookaheadType !== $token && ($token !== Lexer::T_IDENTIFIER || $lookaheadType <= Lexer::T_IDENTIFIER)) {
```
Isn't the second check useless? `$lookaheadType <= Lexer::T_IDENTIFIER` is always true if `$token !== Lexer::T_IDENTIFIER`

Ocramius commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -949,26 +953,34 @@ public function AliasIdentificationVariable()
      * @return string
      */
-    public function AbstractSchemaName()
+    public function AbstractSchemaName($validateName = true)
```
Validating that the class exists seems to be a separate concern to me. Why did you introduce it?

mpdude replied:
Parser::NewObjectExpression does not want validation, at least not immediately when reading the AbstractSchemaName.

I could also extract the validation into a separate method but then you'd need to call them in several places and – most importantly – must not forget to call it when you use AbstractSchemaName. Having the validation toggleable via a parameter is ugly but errs on the side of applying too much validation.

You choose.

Ocramius commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -1125,15 +1125,18 @@ public function walkJoin($join)
                 $dqlAlias   = $joinDeclaration->aliasIdentificationVariable;
                 $tableAlias = $this->getSQLTableAlias($class->table['name'], $dqlAlias);
-                $condition = '(' . $this->walkConditionalExpression($join->conditionalExpression) . ')';
+                $conditions = array();
```
You can safely use `[]` ;-)

Ocramius commented on tests/Doctrine/Tests/ORM/Query/ParserTest.php:
```diff
@@ -0,0 +1,145 @@
+{
+
+    /**
+     * @covers \Doctrine\ORM\Query\Parser::AbstractSchemaName
```
You can simply use `@covers \Doctrine\ORM\Query\Parser` at class level ;-)

mpdude replied:
I wanted to even limit it to the method level to make sure I've got relevant parts covered.

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -462,7 +490,7 @@ public function testSupportsInstanceOfExpressionsInWherePart()
     {
         $this->assertSqlGeneration(
-            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF (Doctrine\Tests\Models\Company\CompanyEmployee, \Doctrine\Tests\Models\Company\CompanyManager)",
+            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF (Doctrine\Tests\Models\Company\CompanyEmployee, Doctrine\Tests\Models\Company\CompanyManager)",
```
Can this test be reverted?

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -473,7 +501,7 @@ public function testSupportsInstanceOfExpressionInWherePartWithMultipleValues()
     {
         $this->assertSqlGeneration(
-            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF \Doctrine\Tests\Models\Company\CompanyEmployee",
+            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF Doctrine\Tests\Models\Company\CompanyEmployee",
```
Can this change be reverted?

Ocramius commented on tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php:
```diff
@@ -484,7 +512,7 @@ public function testSupportsInstanceOfExpressionsInWherePartPrefixedSlash()
     {
         $this->assertInvalidSqlGeneration(
-            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF \Doctrine\Tests\Models\CMS\CmsUser",
+            "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF Doctrine\Tests\Models\CMS\CmsUser",
```
Can this change be reverted?

Ocramius commented:
@mpdude overall big :+1: for the improvements around this PR in general, as we now have some decent coverage of this logic.

Missing bits to change would be in the EBNF documentation ( http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html#ebnf - https://github.com/doctrine/doctrine2/blob/b055d78ea19835fab563dfc234d4804a9d04966a/docs/en/reference/dql-doctrine-query-language.rst )

mpdude committed:
Revert unnecessary changes, minor CS tweaks

mpdude committed:
Use Lexer::T_FULLY_QUALIFIED_NAME instead of T_QUALIFIED_NAME

mpdude committed:
Update the BNF docs

mpdude committed:
Extract method for validating AbstractSchemaNames

mpdude commented:
@Ocramius Thanks! I followed all your comments.

Ocramius commented:
@mpdude mpdude/doctrine2@033b473 broke the build though :-|

mpdude committed:
Remove tests that checked whether AbstractSchemaName validated class names

mpdude commented:
Ooops, forgot to remove the test that checks for the classname validation in AbstractSchemaName.

guilhermeblanco commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -1443,8 +1445,8 @@ Identifiers
-    /* identifier that must be a class name (the "User" of "FROM User u") */
-    AbstractSchemaName ::= identifier
+    /* identifier that must be a class name (the "User" of "FROM User u"), possibly as a fully qualified class name or namespace-aliased */
+    AbstractSchemaName ::= fully_qualified_name | aliased_name | identifier
```
`fully_qualified_name` and `aliased_name` are not terminals. Keep the `identifier` (which holds both of your conceptualized values) only

mpdude replied:
Might get you wrong but I think this PR just changed that...?

guilhermeblanco commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -126,10 +128,11 @@ public function __construct($input)
-            "'(?:[^']|'')*'",
-            '\?[0-9]*|:[a-z_][a-z0-9_]*'
+            '[a-z_][a-z0-9_]*\:[a-z_][a-z0-9_]*', // aliased name
+            '\\\?[a-z_][a-z0-9_]*(?:\\\[a-z_][a-z0-9_]*)*', // identifier or qualified name
```
We don't want to handle user mistakes here. Please remove the FQCN prefixed by `\` support.

mpdude replied:
It was intentionally added a while ago, see https://github.com/doctrine/doctrine2/pull/1395#issuecomment-142230552

guilhermeblanco replied:
It was added because early adopters (thus the low value in DDC ticket) used `\` a lot. We never wanted that though. It gets stripped before even sent anywhere. In DQL things are different, you gotta send it the proper value. =)

mpdude replied:
So you definetly don't want it in the `FROM` clause.

What about `INSTANCE OF`?

guilhermeblanco commented on lib/Doctrine/ORM/Query/Lexer.php:
```diff
@@ -126,10 +128,11 @@ public function __construct($input)
-            '(?:[0-9]+(?:[\.][0-9]+)*)(?:e[+-]?[0-9]+)?',
-            "'(?:[^']|'')*'",
-            '\?[0-9]*|:[a-z_][a-z0-9_]*'
+            '[a-z_][a-z0-9_]*\:[a-z_][a-z0-9_]*', // aliased name
```
I don't understand why you're separating them. You're increasing the complexity of the regex (which is already not so simple) for - what I can see so far, top-to-bottom reviewing - no reason.

mpdude replied:
Ok, let me try to explain.

Previously, `user`, `Alias:Class` or `Some\Class\Name` were all considered "identifiers". This is in the `'[a-z_\\\][a-z0-9_\:\\\]*[a-z0-9_]{1}'` regexp.

Now there are several parts in the DQL grammar where these `identifiers` were accepted, e. g. `AbstractSchemaName`, `IdentificationVariable` or `AliasIdentificationVariable`. So was possible to use a FQCN the the `IdentificationVariable` and you would not get parser complaints (only other failures further downstream).

At first, I tried to add the `\` and `:` as tokens and then parse `AbstractSchemaName` as a combination of repeated `identifier`, `\` and `:`. In principle, that worked, with the only problem that for example `FROM \Some\ Class` would *also* be accepted. This is due to the way the tokenizer works: It would deliver `FROM`, `\`, `Some`(identifier), `\`, `Class` (identifier). 

So, the only solution I saw was to treat `identifier`, `aliased name` and `fully qualified name` as different tokens at the Lexer level already.

Does that make sense?

guilhermeblanco commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -949,29 +954,43 @@ public function AliasIdentificationVariable()
-            list($namespaceAlias, $simpleClassName) = explode(':', $schemaName);
-
-            $schemaName = $this->em->getConfiguration()->getEntityNamespace($namespaceAlias) . '\\' . $simpleClassName;
+        if ($this->lexer->isNextToken(Lexer::T_FULLY_QUALIFIED_NAME)) {
```
Sorry if my review sounds like bashing your efforts, but here we're breaking BC.
Currenlty we support users to do: `AppBundle:Namespaced\MyEntity` which would now not be supported anymore (because the regex rule would reject it).

Ocramius replied:
That's actually what I though. So we should handle also `Foo:Bar\Baz`?

mpdude replied:
@guilhermeblanco Don't worry, I am glad you find the time to review this. This is definetly too big for my brain alone.

This should not be a showstopper, you guys just have to decide how it should work.

mpdude replied:
Fixed this BC break

guilhermeblanco commented on lib/Doctrine/ORM/Query/Parser.php:
```diff
@@ -1811,24 +1832,16 @@ public function PartialObjectExpression()
-            $className = $this->em->getConfiguration()
-                ->getEntityNamespace($namespaceAlias) . '\\' . $simpleClassName;
-        }
+        $className = $this->AbstractSchemaName(); // note that this is not yet validated
```
Is there a specific reason why we need to decouple the validation from the token parser itself?
Is there a reason why we can't validate this here?

mpdude replied:
I don't know. It was like this before, only that I switched from expecting an `identifier` token here to an `AbstractSchemaName` (and I had to decouple the latter from the validation to keep things as they were previously).

guilhermeblanco commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -1125,15 +1125,18 @@ public function walkJoin($join)
-                $condition = '(' . $this->walkConditionalExpression($join->conditionalExpression) . ')';
+                $conditions = [];
+
+                if ($join->conditionalExpression) {
```
Looks like a valid fix, but unrelated to this PR.
From what I could quickly see, you're handling the dis-jointed class support, such as: `SELECT u FROM User u JOIN Address a ON a.user_id = u.id`

Ocramius replied:
@guilhermeblanco the PR is about two tickets

mpdude replied:
@guilhermeblanco I was hunting a bug with that in the first place and then discovered that the parser does not implement `match()` correctly (which IMO would have made my original problem surface earlier).

I think I have provided some tests for that specific change, probably also described in #1328 and #1194.

mpdude committed:
Support aliased + partially namespaced names like AppBundle:Namespaced\MyEntity

mpdude committed:
No longer allow leading backslashes in FQCNs

See https://github.com/doctrine/doctrine2/pull/1395#discussion-diff-40799549

mpdude committed:
Fix some more tests because now a leading backslash in front of class names is disallowed

mpdude commented:
Tests are green now.

guilhermeblanco commented:
Closed in favor of #1549.

Thanks a lot for your help! It's not everyone that manages to dig so deeply and understand DQL parser and you've done an amazing job working on it. In the follow-up PR I basically re-introduced backslash to not create a BC and I fixed a couple hidden bugs I've found while traversing through you patch.

I'm very confident you now have a good understanding of how DQL parser works, so I'll explain which changes I've done, so you're aware.
Keywords had to be separate from identifiers. This was required because isNextToken() applied an exact match, while match() itself was very relaxed. That means match(FROM) was valid even though it was not FROM itself, because it could be considered as an identifier.
To address it, I updated your changed condition in match() and I separated non-terminals (0-99), identifiers (100-199) and keyword (200+). This helped me to make things more strict.
